modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest.geoip;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.VersionedNamedWriteable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.persistent.PersistentTaskState;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.ingest.geoip.GeoIpDownloader.GEOIP_DOWNLOADER;
+
+public class GeoIpTaskState implements PersistentTaskState, VersionedNamedWriteable {
+
+    private static ParseField DATABASES = new ParseField("databases");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<GeoIpTaskState, Void> PARSER =
+        new ConstructingObjectParser<>(GEOIP_DOWNLOADER, true,
+            args -> {
+                List<Map<String, Metadata>> databases = (List<Map<String, Metadata>>) args[0];
+                GeoIpTaskState geoIpTaskState = new GeoIpTaskState();
+                for (Map<String, Metadata> database : databases) {
+                    geoIpTaskState.databases.putAll(database);
+                }
+                return geoIpTaskState;
+            });
+
+    static {
+        PARSER.declareNamedObjects(constructorArg(), (p, c, name) -> Map.of(name, Metadata.fromXContent(p)), DATABASES);
+    }
+
+    public static GeoIpTaskState fromXContent(XContentParser parser) {
+        try {
+            return PARSER.parse(parser, null);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final Map<String, Metadata> databases = new TreeMap<>();
+
+    public GeoIpTaskState() {
```

Comment:
I think these constructors can be package protected?

Reply:
I reworked `GeoIpTaskState` to make it immutable, this constructor is gone now